cpp/src/arrow/compute/exec/hash_join.h

```diff
@@ -56,19 +56,36 @@ class ARROW_EXPORT HashJoinSchema {
                                 const std::string& left_field_name_prefix,
                                 const std::string& right_field_name_prefix);
 
+  Result<Expression> BindFilter(Expression filter, const Schema& left_schema,
+                                const Schema& right_schema);
   std::shared_ptr<Schema> MakeOutputSchema(const std::string& left_field_name_prefix,
                                            const std::string& right_field_name_prefix);
 
+  bool HasLeftPayload() { return HasPayload(0); }
```

Comment:
Nit: could you replace this with `LeftPayloadIsEmpty` for greater specificity?

Reply:
Actually thinking about this one more too, I think it would be better to leave it as some sort of positive, as the code paths that use this all check if a payload exists, not if a payload doesn't exist. A different name could be "LeftPayloadExists" or "LeftPayloadIsNonempty".